java/ql/src/utils/stub-generator/Stubs.qll

```diff
@@ -285,14 +285,19 @@ private string stubQualifier(RefType t) {
     else result = ""
 }
 
+pragma[nomagic]
+private predicate needsPackageNameHelper(RefType t, GeneratedTopLevel top, string name) {
+  t.getSourceDeclaration() = [getAReferencedType(top), top].getSourceDeclaration() and
+  name = t.getName()
+}
+
 /**
  * Holds if `t` may clash with another type of the same name, so should be referred to using the fully qualified name
  */
 private predicate needsPackageName(RefType t) {
-  exists(GeneratedTopLevel top, RefType other |
-    t.getSourceDeclaration() = [getAReferencedType(top), top].getSourceDeclaration() and
-    other.getSourceDeclaration() = [getAReferencedType(top), top].getSourceDeclaration() and
-    t.getName() = other.getName() and
+  exists(GeneratedTopLevel top, RefType other, string name |
+    needsPackageNameHelper(t, top, name) and
+    needsPackageNameHelper(other, top, name) and
     t != other
```

Comment:
Does it perform better, if you replace this with
```ql
not exists(unique(RefType t0 | needsPackageNameHelper(t0, top, name) | t0))
```
?

Reply:
I expect not, since you'll still put the combination of `t` and `t0` in the same scope, which causes a quadratic combination. Instead this property of having multiple types with a given pair of `top, name` can be factored into it's own predicate without putting two `RefType`s in scope.
```
pragma[nomagic]
private predicate describesMultipleTypes(GeneratedTopLevel top, string name) {
  2 <= strictcount(RefType t | needsPackageNameHelper(t, top, name))
}

private predicate needsPackageName(RefType t) {
  exists(GeneratedTopLevel top, string name |
    needsPackageNameHelper(t, top, name) and
    describesMultipleTypes(top, name)
  )
}
```

Reply:
> Does it perform better, if you replace these two lines with
> 
> ```ql
> not exists(unique(RefType t0 | needsPackageNameHelper(t0, top, name) | t0))
> ```
> 
> ?

It does a bit better, but `needsPackageNameHelper` seems to still have some bad join order:

```
[2023-05-26 16:38:58] (14s) Tuple counts for Stubs#f198c315::needsPackageNameHelper#3#fff/3@de488e5i after 10.2s:
                      112757    ~0%       {3} r1 = JOIN Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff WITH Element#c53922e0::Element::hasName#1#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.0 't', Rhs.1 'name'
                      437133995 ~4%       {3} r2 = JOIN r1 WITH Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 'top', Lhs.1 't', Lhs.2 'name'
                      
                      102929    ~5%       {3} r3 = JOIN r2 WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1 't', Lhs.0 'top', Lhs.2 'name'
                      
                      12709047  ~353%     {3} r4 = JOIN r2 WITH Stubs#f198c315::getAReferencedType#1#bf_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 'top', Lhs.1 't', Lhs.2 'name'
                      4910018   ~125%     {3} r5 = JOIN r4 WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1 't', Lhs.0 'top', Lhs.2 'name'
                      
                      5012947   ~128%     {3} r6 = r3 UNION r5
                                          return r6
```



> I expect not, since you'll still put the combination of `t` and `t0` in the same scope, which causes a quadratic combination. Instead this property of having multiple types with a given pair of `top, name` can be factored into it's own predicate without putting two `RefType`s in scope.
> 
> ```
> pragma[nomagic]
> private predicate describesMultipleTypes(GeneratedTopLevel top, string name) {
>   2 <= strictcount(RefType t | needsPackageNameHelper(t, top, name))
> }
> 
> private predicate needsPackageName(RefType t) {
>   exists(GeneratedTopLevel top, string name |
>     needsPackageNameHelper(t, top, name) and
>     describesMultipleTypes(top, name)
>   )
> }
> ```

With this solution, `describesMultipleTypes` finishes very fast, but the tuple counts in `needsPackageNameHelper` remain the same (probably unsurprisingly):

```
[2023-05-26 16:43:08] (13s) Tuple counts for Stubs#f198c315::needsPackageNameHelper#3#fff/3@de488e5i after 9.4s:
                      112757    ~0%       {3} r1 = JOIN Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff WITH Element#c53922e0::Element::hasName#1#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.0 't', Rhs.1 'name'
                      437133995 ~4%       {3} r2 = JOIN r1 WITH Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 'top', Lhs.1 't', Lhs.2 'name'
                      
                      102929    ~5%       {3} r3 = JOIN r2 WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1 't', Lhs.0 'top', Lhs.2 'name'
                      
                      12709047  ~353%     {3} r4 = JOIN r2 WITH Stubs#f198c315::getAReferencedType#1#bf_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 'top', Lhs.1 't', Lhs.2 'name'
                      4910018   ~125%     {3} r5 = JOIN r4 WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1 't', Lhs.0 'top', Lhs.2 'name'
                      
                      5012947   ~128%     {3} r6 = r3 UNION r5
                                          return r6
```

Reply:
With:

```ql
pragma[nomagic]
private predicate needsPackageNameHelper(RefType t, GeneratedTopLevel top, string name) {
  t.getSourceDeclaration() =
    pragma[only_bind_out]([getAReferencedType(top), top].getSourceDeclaration()) and
  name = t.getName()
}

```

It looks much better:

```
[2023-05-29 09:55:20] (8s) Tuple counts for Stubs#f198c315::needsPackageNameHelper#3#fff/3@9989262i after 902ms:
                      426     ~1%      {2} r1 = JOIN Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0 'top'
                      
                      21456   ~0%      {2} r2 = SCAN Stubs#f198c315::getAReferencedType#1#bf OUTPUT In.1, In.0 'top'
                      21456   ~49%     {2} r3 = JOIN r2 WITH Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1 'top', Rhs.1
                      11903   ~3%      {2} r4 = JOIN r3 WITH Stubs#f198c315::GeneratedTopLevel#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0 'top'
                      
                      12329   ~5%      {2} r5 = r1 UNION r4
                      2470396 ~1%      {2} r6 = JOIN r5 WITH Type#6144c3fd::RefType::getSourceDeclaration#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 't', Lhs.1 'top'
                      2470396 ~13%     {3} r7 = JOIN r6 WITH Element#c53922e0::Element::hasName#1#dispred#ff ON FIRST 1 OUTPUT Lhs.0 't', Lhs.1 'top', Rhs.1 'name'
                                       return r7
```

And there are no significantly slow predicates:

```
Most expensive predicates for completed query MinimalStubsFromSource.ql:
        time  | evals |   max @ iter | predicate
        ------|-------|--------------|----------
        903ms |       |              | Stubs#f198c315::needsPackageNameHelper#3#fff@9989262i
        580ms |    14 | 218ms @ 2    | __Generics#716e7108::GenericType#f_Generics#716e7108::ParameterizedType#f_Generics#716e7108::Paramet__#join_rhs@L5#fcd8b
        491ms |       |              | Stubs#f198c315::needsPackageName#1#f@1f8717nr
        489ms |       |              | Stubs#f198c315::needsPackageNameHelper#3#fff_120#count_range@402a3f2g
        342ms |    14 | 124ms @ 2    | Stubs#f198c315::stubGenericArguments#2#fbf@fcd8bw6o
        297ms |       |              | ___files_locations_default#shared_hasLocation_10#join_rhs#min_range___files_locations_default#shared__#join_rhs@099fbb03
        284ms |     8 | 139ms @ 3    | Stubs#f198c315::getAContainedType#1#ff@4eb28abu
...
```